### PR TITLE
test(scorecard): add custom target discovery to recording scorecard test

### PR DESCRIPTION
**Pull request by mwangggg**
# Welcome to Cryostat! 👋
## Before contributing, make sure you have:
* [x] Read the [contributing guidelines](https://github.com/cryostatio/cryostat/blob/main/CONTRIBUTING.md)
* [x] Linked a relevant issue which this PR resolves
* [x] Linked any other relevant issues, PR's, or documentation, if any
* [x] Resolved all conflicts, if any
* [x] Rebased your branch PR on top of the latest upstream `main` branch
* [x] Attached at least one of the following labels to the PR: `[chore, ci, docs, feat, fix, test]`
* [x] [Signed all commits](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits): `git commit -S -m "YOUR_COMMIT_MESSAGE"`
_______________________________________________

Related to: https://github.com/cryostatio/cryostat-operator/issues/504

### CODE REVIEW

**Comment by mwangggg** [comment]
oh I guess I can't do that here..

**Comment by andrewazores** [comment]
Looks reasonable to me.

**Comment by ebaron** [comment]
This has been passing much more reliably for me than with the built-in discovery. Although about 1/10th of the time I'll get the following EOF:

```
 --------------------------------------------------------------------------------
 Image:      quay.io/ebaron/cryostat-operator-scorecard:mwangggg-698-rec-scorecard-cont-01
 Entrypoint: [cryostat-scorecard-tests cryostat-recording]
 Labels:
     "suite":"cryostat"
     "test":"cryostat-recording"
 Results:
     Name: cryostat-recording
     State: fail

     Errors:
         failed to archive the recording: Patch "https://cryostat-recording-cryostat-operator-scorecard.apps.example.com/api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Flocalhost:0%2Fjmxrmi/recordings/scorecard_test_rec": EOF
     Log:
         deployment cryostat-recording is not yet found
         deployment cryostat-recording is not yet found
         deployment cryostat-recording is not yet found
         deployment cryostat-recording is not yet found
         deployment cryostat-recording is not yet found
         deployment cryostat-recording is not yet found
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is not yet available
         deployment cryostat-recording is available
         application is available at https://cryostat-recording-cryostat-operator-scorecard.apps.example.com
         application is not yet reachable at https://cryostat-recording-cryostat-operator-scorecard.apps.example.com
         application is not yet reachable at https://cryostat-recording-cryostat-operator-scorecard.apps.example.com
         application is not yet reachable at https://cryostat-recording-cryostat-operator-scorecard.apps.example.com
         application is not yet reachable at https://cryostat-recording-cryostat-operator-scorecard.apps.example.com
         application is ready at https://cryostat-recording-cryostat-operator-scorecard.apps.example.com
         created a custom target: &{ConnectUrl:service:jmx:rmi:///jndi/rmi://localhost:0/jmxrmi Alias:customTarget}
         created stored credential with match expression: target.alias=="customTarget"
         created a recording: &{DownloadURL:https://cryostat-recording-cryostat-operator-scorecard.apps.example.com:443/api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Flocalhost:0%2Fjmxrmi/recordings/scorecard_test_rec ReportURL:https://cryostat-recording-cryostat-operator-scorecard.apps.example.com:443/api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Flocalhost:0%2Fjmxrmi/reports/scorecard_test_rec Id:1 Name:scorecard_test_rec StartTime:1709136597936 State:RUNNING Duration:0 Continuous:true ToDisk:true MaxSize:0 MaxAge:0}
         current list of recordings: [{DownloadURL:https://cryostat-recording-cryostat-operator-scorecard.apps.8a7548441a03cb1654f8.hypershift.aws-2.ci.openshift.org:443/api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Flocalhost:0%2Fjmxrmi/recordings/scorecard_test_rec ReportURL:https://cryostat-recording-cryostat-operator-scorecard.apps.example.com:443/api/v1/targets/service:jmx:rmi:%2F%2F%2Fjndi%2Frmi:%2F%2Flocalhost:0%2Fjmxrmi/reports/scorecard_test_rec Id:1 Name:scorecard_test_rec StartTime:1709136597936 State:RUNNING Duration:0 Continuous:true ToDisk:true MaxSize:0 MaxAge:0}]
```

This seems a bit similar to the error we encountered with the built-in discovery. There are some things we can try on the Golang client side, like: https://stackoverflow.com/a/34474535. I'm not sure if that's what we're seeing though. Given that the tests seem more reliable now, I think we should get the tests merged and investigate further.

File: internal/test/scorecard/clients.go [hunk 1]
```diff
@@ -242,9 +243,36 @@ func (client *TargetClient) List(ctx context.Context) ([]Target, error) {
...
+	}
 	defer resp.Body.Close()
 
-	return targets, nil
+	if !StatusOK(resp.StatusCode) {
+		return nil, fmt.Errorf("API request failed with status code %d: %s", resp.StatusCode, ReadError(resp))
+	}
+
+	targetResp := &CustomTargetResponse{}
+	err = ReadJSON(resp, &targetResp)
```
Comment (tthvo) [inline internal/test/scorecard/clients.go]:
```suggestion
	err = ReadJSON(resp, targetResp)
```

`targetResp` is already a pointer right?

**Review by tthvo — COMMENTED** [review]
> This seems a bit similar to the error we encountered with the built-in discovery. There are some things we can try on the Golang client side, like: https://stackoverflow.com/a/34474535. I'm not sure if that's what we're seeing though. Given that the tests seem more reliable now, I think we should get the tests merged and investigate further.

Hey @ebaron  Looks like some rate limit issue (not sure though)?. Maybe another option can be a retry when EOF is encountered?

Something like a `wait` loop with a check for `errors.Is(err, io.EOF)`?

**Comment by ebaron** [comment]
> > This seems a bit similar to the error we encountered with the built-in discovery. There are some things we can try on the Golang client side, like: https://stackoverflow.com/a/34474535. I'm not sure if that's what we're seeing though. Given that the tests seem more reliable now, I think we should get the tests merged and investigate further.
> 
> Hey @ebaron Looks like some rate limit issue (not sure though)?. Maybe another option can be a retry when EOF is encountered?
> 
> Something like a `wait` loop with a check for `errors.Is(err, io.EOF)`?

That sounds like a good idea!

**Comment by andrewazores** [comment]
Would it also be possible to dump out the response status code and headers when there's a failure? That could also help a lot with troubleshooting - it could tell us what the server thinks went wrong, or if it is some kind of rate-limiting behaviour then we should be able to tell that the response actually came from the Service/Ingress/etc., too.

**Comment by tthvo** [comment]
> Would it also be possible to dump out the response status code and headers when there's a failure? That could also help a lot with troubleshooting - it could tell us what the server thinks went wrong, or if it is some kind of rate-limiting behaviour then we should be able to tell that the response actually came from the Service/Ingress/etc., too.

Sure good idea! The test currently emits a status code and response body on non-200 code. I can also log out headers too.

https://pkg.go.dev/net/http#Response

**Review by tthvo — APPROVED** [review]
Changes look good to me! We will note the ideas/suggestions above down and continue on upstream!

**Pull request merged**